localization/ndt_scan_matcher/src/diagnostics_module.cpp

```diff
@@ -0,0 +1,137 @@
+  diagnostics_status_msg_.hardware_id = node_->get_name();
+}
+
+void DiagnosticsModule::clear()
+{
+  clearKeyValue();
+  clearLevelAndMessage();
+}
+
+void DiagnosticsModule::clearKeyValue()
+{
+  diagnostics_status_msg_.values.clear();
+  diagnostics_status_msg_.values.shrink_to_fit();
+}
+
+void DiagnosticsModule::clearLevelAndMessage()
+{
+  diagnostics_status_msg_.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+  diagnostics_status_msg_.message = "";
+}
+
+void DiagnosticsModule::addKeyValue(const diagnostic_msgs::msg::KeyValue & key_value_msg)
+{
+  auto it = std::find_if(
+    std::begin(diagnostics_status_msg_.values), std::end(diagnostics_status_msg_.values),
+    [key_value_msg](const auto & arg) { return arg.key == key_value_msg.key; });
+
+  if (it != std::cend(diagnostics_status_msg_.values)) {
+    it->value = key_value_msg.value;
+  } else {
+    diagnostics_status_msg_.values.push_back(key_value_msg);
+  }
+}
+
+template <>
+void DiagnosticsModule::addKeyValue(const std::string & key, const std::string & value)
+{
+  diagnostic_msgs::msg::KeyValue key_value;
+  key_value.key = key;
+  key_value.value = value;
+  addKeyValue(key_value);
+}
+
+template <>
+void DiagnosticsModule::addKeyValue(const std::string & key, const bool & value)
+{
+  diagnostic_msgs::msg::KeyValue key_value;
+  key_value.key = key;
+  key_value.value = value ? "True" : "False";
+  addKeyValue(key_value);
+}
+
+void DiagnosticsModule::updateLevelAndMessage(const int8_t level, const std::string & message)
+{
+  if ((level > diagnostic_msgs::msg::DiagnosticStatus::OK)) {
+    if (!diagnostics_status_msg_.message.empty()) {
+      diagnostics_status_msg_.message += "; ";
+    }
+    diagnostics_status_msg_.message += message;
+  }
+  if (level > diagnostics_status_msg_.level) {
+    diagnostics_status_msg_.level = level;
+  }
+}
+
+void DiagnosticsModule::publish()
+{
+  diagnostics_pub_->publish(createDiagnosticsArray());
+}
+
+diagnostic_msgs::msg::DiagnosticArray DiagnosticsModule::createDiagnosticsArray() const
+{
+  diagnostic_msgs::msg::DiagnosticArray diagnostics_msg;
+  diagnostics_msg.header.stamp = node_->now();
+  diagnostics_msg.status.push_back(diagnostics_status_msg_);
+
+  if (diagnostics_msg.status.at(0).level == diagnostic_msgs::msg::DiagnosticStatus::OK) {
+    diagnostics_msg.status.at(0).message = "OK";
+  }
+
+  return diagnostics_msg;
+}
+
+std::vector<diagnostic_msgs::msg::KeyValue>::const_iterator DiagnosticsModule::findIteratorByKey(
+  const std::string & key) const
+{
+  const auto it = std::find_if(
+    std::begin(diagnostics_status_msg_.values), std::end(diagnostics_status_msg_.values),
+    [key](const auto & arg) { return arg.key == key; });
+  return it;
+}
+
+bool DiagnosticsModule::existIterator(
+  const std::vector<diagnostic_msgs::msg::KeyValue>::const_iterator & it) const
+{
+  return it != std::cend(diagnostics_status_msg_.values);
+}
```

Comment:
I think the following changes made these codes unnecessary.
https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/f975f1bee7ef607fe1d545859db45a541f43ef4a

Reply:
fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/3cd9471c9952010916a18064ca83561971d156a7